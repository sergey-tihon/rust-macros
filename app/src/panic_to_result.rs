use panic_to_result_macro::panic_to_result;

#[allow(dead_code)]
#[derive(Debug)]
pub struct Person {
    name: String,
    age: u8,
}

#[panic_to_result]
fn create_person(name: String, age: u8) -> Person {
    if age < 18 {
        panic!("Age must be 18 or older");
    }
    Person { name, age }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn happy_path() {
        let actual = create_person("Alice".to_string(), 22).unwrap();

        assert_eq!(actual.name, "Alice".to_string());
        assert_eq!(actual.age, 22);
    }

    #[test]
    fn should_panic_on_invalid_age() {
        let actual = create_person("Alice".to_string(), 17);

        assert_eq!(
            actual.expect_err("this should be an error"),
            "Age must be 18 or older"
        );
    }
}
