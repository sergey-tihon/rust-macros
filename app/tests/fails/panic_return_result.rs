use panic_to_result_macro::panic_to_result;

#[derive(Debug)]
pub struct Person {
    name: String,
    age: u8,
}

#[panic_to_result]
fn create_person(name: String, age: u8) -> Result<Person, String> {
    if age < 18 {
        panic!("Age must be 18 or older");
    }
    Ok(Person { name, age })
}

fn main() {}
