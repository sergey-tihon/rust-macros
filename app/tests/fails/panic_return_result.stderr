error: this macro can only be applied to a function that does not return a Result. Signature: Result < Person, String >
  --> tests/fails/panic_return_result.rs:10:44
   |
10 | fn create_person(name: String, age: u8) -> Result<Person, String> {
   |                                            ^^^^^^^^^^^^^^^^^^^^^^
